module/zfs/dmu_recv.c

```diff
@@ -1207,6 +1257,261 @@ dmu_recv_begin(char *tofs, char *tosnap, dmu_replay_record_t *drr_begin,
+			dsl_dataset_t *ds;
+			dsl_crypto_key_t *dck = NULL;
+			uint8_t salt[ZIO_DATA_SALT_LEN];
+			uint8_t iv[ZIO_DATA_IV_LEN];
+			uint8_t mac[ZIO_DATA_MAC_LEN];
+			boolean_t no_crypt = B_FALSE;
+			dsl_pool_t *dp =
+			    rwa->os->os_dsl_dataset->ds_dir->dd_pool;
+			enc_buf = zio_data_buf_alloc(lsize);
+			abd_t *eabd = abd_get_from_buf(enc_buf, lsize);
+			abd_take_ownership_of_buf(eabd, B_FALSE);
+			zio_crypt_decode_params_bp(bp, salt, iv);
+			zio_crypt_decode_mac_bp(bp, mac);
+
+			dsl_pool_config_enter(dp, FTAG);
+			err = dsl_dataset_hold_flags(dp, rwa->tofs,
+			    DS_HOLD_FLAG_DECRYPT, FTAG, &ds);
+			if (err != 0) {
+				dsl_pool_config_exit(dp, FTAG);
+				abd_release_ownership_of_buf(abd);
+				abd_put(abd);
+				if (buf != NULL)
+					abd_free(abd);
+				abd_release_ownership_of_buf(eabd);
+				abd_put(eabd);
+				abd_free(eabd);
+				zio_data_buf_free(enc_buf, lsize);
+				return (SET_ERROR(EACCES));
+			}
+
+			/* Look up the key from the spa's keystore */
+			err = spa_keystore_lookup_key(rwa->os->os_spa,
+			    zb.zb_objset, FTAG, &dck);
+			if (err != 0) {
+				dsl_dataset_rele_flags(ds, DS_HOLD_FLAG_DECRYPT,
+				    FTAG);
+				dsl_pool_config_exit(dp, FTAG);
+				abd_release_ownership_of_buf(abd);
+				abd_put(abd);
+				if (buf != NULL)
+					abd_free(abd);
+				abd_release_ownership_of_buf(eabd);
+				abd_put(eabd);
+				abd_free(eabd);
+				zio_data_buf_free(enc_buf, lsize);
+				return (SET_ERROR(EACCES));
+			}
+
+			err = zio_do_crypt_abd(B_TRUE,  &dck->dck_key,
+			    BP_GET_TYPE(bp), BP_SHOULD_BYTESWAP(bp), salt, iv,
+			    mac, lsize, abd, eabd, &no_crypt);
+
+			spa_keystore_dsl_key_rele(rwa->os->os_spa, dck, FTAG);
+			dsl_dataset_rele_flags(ds, DS_HOLD_FLAG_DECRYPT, FTAG);
+			dsl_pool_config_exit(dp, FTAG);
+
+			if (err != 0 || no_crypt) {
+				abd_release_ownership_of_buf(abd);
+				abd_put(abd);
+				if (buf != NULL)
+					abd_free(abd);
+				abd_release_ownership_of_buf(eabd);
+				abd_put(eabd);
+				abd_free(eabd);
+				zio_data_buf_free(enc_buf, lsize);
+				return (err == 0 ? SET_ERROR(ENOTSUP) : err);
+			}
+
+			/* Swap in the newly encrypted data into the abd */
+			abd_release_ownership_of_buf(abd);
+			abd_put(abd);
+			if (buf != NULL) {
+				abd_free(abd);
+				buf = NULL;
+			}
+			abd = eabd;
+
+			/*
+			 * We want to prevent zio_rewrite() from trying to
+			 * encrypt the data again
+			 */
+			flags |= ZIO_FLAG_RAW;
+		}
+
+		if (BP_GET_COMPRESS(bp) != ZIO_COMPRESS_OFF) {
+			/* Recompress the data */
+			if (buf != NULL)
+				abd_free(abd);
+			size = zio_compress_data(BP_GET_COMPRESS(bp), abd, buf,
+			    lsize);
+		}
+	}
+
+	io = zio_rewrite(NULL, rwa->os->os_spa, 0, bp, abd,
+	    size, NULL, NULL, ZIO_PRIORITY_SYNC_WRITE, flags, &zb);
+
+	/* compute new bp checksum value and make sure it matches the old one */
+	zio_checksum_compute(io, BP_GET_CHECKSUM(bp), abd, size);
+	if (size != BP_GET_PSIZE(bp) ||
+	    !ZIO_CHECKSUM_EQUAL(bp_cksum, io->io_bp->blk_cksum)) {
```

Comment:
this might be simplified by using `zio_checksum_error_impl()`, in which case you could verify the checksum before creation the zio, and not have to save the bp_cksum.

Reply:
I'm not sure that using this rather complicate function is simpler than stashing the BP and using the macro